### PR TITLE
chore: extract sink forwarder out of the pkg/forwarder

### PR DESCRIPTION
**Pull request by jy4096**
- fix source pkg import ordering
- add sink forwarder
- remove terminal applier because no one uses it after we extract the sink forwarder

fixes https://github.com/numaproj/numaflow/issues/985

TODO (in a different PR): revisit idle handler implementation when watermark is disabled

### CODE REVIEW

**Comment by jy4096** [comment]
test fail from time to time

```
=== RUN   TestForwarderJetStreamBuffer/batch
{"level":"info","ts":"2023-08-29T00:54:42.719515-07:00","logger":"numaflow","caller":"jetstream/writer.go:112","msg":"Usage is greater than bufferUsageLimit","bufferWriter":"TestForwarderJetStreamBuffer","stream":"TestForwarderJetStreamBuffer","subject":"TestForwarderJetStreamBuffer","partitionIdx":0,"solidUsage":1,"softUsage":1}
{"level":"info","ts":"2023-08-29T00:54:42.722206-07:00","logger":"numaflow","caller":"forward/forward.go:134","msg":"Starting forwarder..."}
{"level":"info","ts":"2023-08-29T00:54:43.724443-07:00","logger":"numaflow","caller":"jetstream/writer.go:112","msg":"Usage is greater than bufferUsageLimit","bufferWriter":"TestForwarderJetStreamBuffer-to","stream":"TestForwarderJetStreamBuffer-to","subject":"TestForwarderJetStreamBuffer-to","partitionIdx":0,"solidUsage":1.1,"softUsage":1.1}
{"level":"info","ts":"2023-08-29T00:54:44.724509-07:00","logger":"numaflow","caller":"jetstream/writer.go:112","msg":"Usage is greater than bufferUsageLimit","bufferWriter":"TestForwarderJetStreamBuffer-to","stream":"TestForwarderJetStreamBuffer-to","subject":"TestForwarderJetStreamBuffer-to","partitionIdx":0,"solidUsage":1.1,"softUsage":1.1}
{"level":"info","ts":"2023-08-29T00:54:45.724537-07:00","logger":"numaflow","caller":"jetstream/writer.go:112","msg":"Usage is greater than bufferUsageLimit","bufferWriter":"TestForwarderJetStreamBuffer-to","stream":"TestForwarderJetStreamBuffer-to","subject":"TestForwarderJetStreamBuffer-to","partitionIdx":0,"solidUsage":1.1,"softUsage":1.1}
{"level":"info","ts":"2023-08-29T00:54:46.724695-07:00","logger":"numaflow","caller":"jetstream/writer.go:112","msg":"Usage is greater than bufferUsageLimit","bufferWriter":"TestForwarderJetStreamBuffer-to","stream":"TestForwarderJetStreamBuffer-to","subject":"TestForwarderJetStreamBuffer-to","partitionIdx":0,"solidUsage":1.1,"softUsage":1.1}
```

**Comment by jy4096** [comment]
another failing unit test
```
{"level":"error","ts":"2023-08-29T21:53:59.987282216Z","logger":"numaflow","caller":"redis/read.go:299","msg":"Stream not found","BufferReader":"somestream","error":"ERR no such key","stacktrace":"github.com/numaproj/numaflow/pkg/isb/stores/redis.(*BufferRead).setError\n\t/home/runner/work/numaflow/numaflow/pkg/isb/stores/redis/read.go:299\ngithub.com/numaproj/numaflow/pkg/isb/stores/redis.(*BufferRead).updateIsEmptyFlag\n\t/home/runner/work/numaflow/numaflow/pkg/isb/stores/redis/read.go:224\ngithub.com/numaproj/numaflow/pkg/isb/stores/redis.(*BufferRead).refreshIsEmptyFlag\n\t/home/runner/work/numaflow/numaflow/pkg/isb/stores/redis/read.go:185"}
{"level":"fatal","ts":"2023-08-29T21:53:59.995053086Z","logger":"numaflow","caller":"testutils/rw.go:120","msg":"context error: context deadline exceeded","stacktrace":"github.com/numaproj/numaflow/pkg/isb/testutils.ReadMessagesLen\n\t/home/runner/work/numaflow/numaflow/pkg/isb/testutils/rw.go:120\ngithub.com/numaproj/numaflow/pkg/isb/stores/redis.TestRedisCheckBacklog\n\t/home/runner/work/numaflow/numaflow/pkg/isb/stores/redis/read_test.go:141\ntesting.tRunner\n\t/opt/hostedtoolcache/go/1.20.7/x64/src/testing/testing.go:1576"}
FAIL	github.com/numaproj/numaflow/pkg/isb/stores/redis	10.046s
```

**Comment by jy4096** [comment]
removing this part triggers the test failure
Also the original code is also not correct?? 
because even though sink vertex doesn't need offet, we still need to check if offset>0 before we publish the watermark (conditional forwarding)
```
else {
	publisher.PublishWatermark(processorWM, nil, int32(index))
	activeWatermarkBuffers[toVertexName][index] = true
	// reset because the toBuffer partition is no longer idling
	isdf.idleManager.Reset(isdf.toBuffers[toVertexName][index].GetName())
}
```

File: pkg/sinks/forward/metrics.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/numaproj/numaflow/pkg/metrics"
+)
+
+// readMessagesCount is used to indicate the number of messages read
+var readMessagesCount = promauto.NewCounterVec(prometheus.CounterOpts{
+	Subsystem: "sink_forwarder",
+	Name:      "read_total",
```
Comment (whynowy) [inline pkg/sinks/forward/metrics.go:29]: Logic in https://github.com/numaproj/numaflow/blob/main/pkg/daemon/server/service/rater/rater.go#L231-L237 has to be updated.

File: pkg/sinks/forward/forward.go [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+
+// DataForward forwards the data from previous vertex to sinker.
+type DataForward struct {
+	ctx context.Context
+	// cancelFn cancels our new context, our cancellation is little more complex and needs to be well orchestrated, hence
+	// we need something more than a cancel().
+	cancelFn            context.CancelFunc
+	fromBufferPartition isb.BufferReader
+	// toBuffers is a map of toVertex name to the toVertex's owned buffers.
+	toBuffers map[string][]isb.BufferWriter
```
Comment (whynowy) [inline pkg/sinks/forward/forward.go:49]: There's only 1 writer, right?

File: pkg/sinks/forward/forward.go [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+	FSD       forward.ToWhichStepDecider
```
Comment (whynowy) [inline pkg/sinks/forward/forward.go:50]: Is it still needed?

File: pkg/sinks/forward/forward.go [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+	wmFetcher fetch.Fetcher
+	// wmPublishers stores the vertex to publisher mapping
+	wmPublishers map[string]publish.Publisher
```
Comment (whynowy) [inline pkg/sinks/forward/forward.go:53]: Should be only 1 as well.

File: pkg/sinks/kafka/kafka.go [hunk 5]
```diff
@@ -83,14 +81,14 @@ func NewToKafka(vertex *dfv1.Vertex,
 	toKafka.topic = kafkaSink.Topic
 	toKafka.kafkaSink = kafkaSink
 
-	forwardOpts := []forward.Option{forward.WithVertexType(dfv1.VertexTypeSink), forward.WithLogger(toKafka.log)}
+	forwardOpts := []sinkforward.Option{sinkforward.WithLogger(toKafka.log)}
```
Comment (whynowy) [inline pkg/sinks/kafka/kafka.go:84]: Do we need to clean up the option `forward.WithVertexType()`?

File: pkg/sinks/forward/forward.go [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+		}
+	}
+	// creating a context here which is managed by the forwarder's lifecycle
+	ctx, cancel := context.WithCancel(context.Background())
+
+	var df = DataForward{
+		ctx:                 ctx,
+		cancelFn:            cancel,
+		fromBufferPartition: fromStep,
+		toBuffer:            toSteps,
```
Comment (whynowy) [inline pkg/sinks/forward/forward.go:82]: nit: `toStep`

File: pkg/sinks/forward/forward.go [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+	if len(writeOffsets) > 0 {
+		df.wmPublisher.PublishWatermark(processorWM, nil, 0)
+	} else {
+		idlehandler.PublishIdleWatermark(ctx, df.toBuffer, df.wmPublisher, df.idleManager, df.opts.logger, dfv1.VertexTypeSink, processorWM)
+	}
+
+	err = df.ackFromBuffer(ctx, readOffsets)
+	// implicit return for posterity :-)
+	if err != nil {
+		df.opts.logger.Errorw("failed to ack from buffer", zap.Error(err))
```
Comment (whynowy) [inline pkg/sinks/forward/forward.go:282]: Nit: `Failed to ack xxxx`

File: pkg/sinks/kafka/kafka.go (hunk 5 above)
Comment (jy4096) [reply pkg/sinks/kafka/kafka.go:84]: yes but redis read test will fail if we remove it

**Comment by jy4096** [comment]
closed by mistake...

**Review by whynowy — APPROVED** [review]

File: pkg/forward/forward.go [hunk 8]
```diff
@@ -350,20 +346,13 @@ func (isdf *InterStepDataForward) forwardAChunk(ctx context.Context) {
 		activeWatermarkBuffers[toVertexName] = make([]bool, len(toVertexBufferOffsets))
 		if publisher, ok := isdf.wmPublishers[toVertexName]; ok {
 			for index, offsets := range toVertexBufferOffsets {
-				if isdf.opts.vertexType == dfv1.VertexTypeMapUDF || isdf.opts.vertexType == dfv1.VertexTypeReduceUDF {
-					if len(offsets) > 0 {
-						publisher.PublishWatermark(processorWM, offsets[len(offsets)-1], int32(index))
-						activeWatermarkBuffers[toVertexName][index] = true
```
[the remark this answers is not in the document]
Comment (vigith) [reply pkg/forward/forward.go:356]: git diff problem :)

**Review by vigith — APPROVED** [review]

**Pull request merged**